### PR TITLE
improve development of evaluations

### LINKED ISSUE #227: Expansion and label appending of output dataset in eval pipeline
After switching to the MetaDataset backend we can end up with lazy datasets which is unintuitive for callbacks. We should `.expand=True` here
https://github.com/pesser/edflow/blob/4cd2225b06c43524c01ab3d512f5d83d10e0a546/edflow/eval/pipeline.py#L278
and here
https://github.com/pesser/edflow/blob/4cd2225b06c43524c01ab3d512f5d83d10e0a546/edflow/eval/pipeline.py#L711
in the latter case, we also don't set `.append_labels` which we do in the first case. I guess this should be made consistent, too.

### PR DESCRIPTION
**Pull request by pesser**
This PR includes some commits that were merged into master directly before switching to the dev branch (regarding EvalPipeline using MetaDataset and something about coveralls).

Fixes #227 and parts of #226.

Besides a bit of cleaning up and fixes, we now have:
- `datasets` dict in config which should be `datasetname: datasetpath`. Results now have the form `datasetname/op/...` which makes it easy to specify on which datasets ops should be executed. This is still a bit experimental but the default works quite well: `train/train_op` is executed every step in train mode, `train/log_op` and `validation/log_op` are logged every `log_freq` steps, `validation/eval_op` runs after each epoch (including callbacks for it, the results of which are logged, too). For backwards compatibility, if no `datasets` dict is found, it is set to `train: dataset, validation: validation_dataset`.
- shortcut for easier resuming and evaluation: `-p <rundir/configs/config.yaml>` translates to `-b <rundir/configs/config.yaml> -p <rundir>`, so you don't have to specify project and config seperately anymore.
- Support for evaluation functors. Running in test mode (without `-t`) and specifying `--eval_functor <keypath>` will init the object under keypath by passing it the config, and replace `validation/eval_op` with the functors `__call__` method, passing `model, **batch` to it. If the functor has a `callbacks` attribute, they will be used.

With the last two points, you can write a simple Functor class to evaluate existing models without changing any code related to that model, eg: `edflow -p <path to config in project folder> --eval_functor <path to functor>`

### CODE REVIEW

File: edflow/main.py [hunk 1]
```diff
@@ -32,110 +17,103 @@ def _save_config(config, prefix="config"):
...
             config["hook_freq"] = 1
         compat_kwargs = dict(
             hook_freq=config["hook_freq"], num_epochs=config["num_epochs"]
         )
         logger.info("Instantiating iterator.")
-        Trainer = implementations["iterator"](
-            config, root, Model, dataset=dataset,
-            validation_dataset=validation_dataset, **compat_kwargs
-        )
+        iterator = iterator(config, root, model, datasets=datasets,
```
Comment (jhaux) [inline edflow/main.py]: Nice! That was long overdue :smile:

File: edflow/custom_logging.py [hunk 2]
```diff
@@ -109,6 +109,8 @@ def init(
                 cls.resumed = False
                 cls.root = os.path.join(log_dir, name)
             else:
+                if run_dir[-1] == "/":
```
Comment (jhaux) [inline edflow/custom_logging.py:112]: This is the same as 
`run_dir = os.path.dirname(run_dir)`
which is potentially more robust.

File: edflow/hooks/logging_hooks/wandb_handler.py [hunk 3]
```diff
@@ -7,8 +7,13 @@ def log_wandb(results, step, path):
...
 
+
 def log_wandb_images(results, step, path):
-    results = dict((
-        path + "/" + k,
-        wandb.Image(((batch_to_canvas(v)+1)*127.5).astype(np.uint8))) for k, v in results.items())
+    results = dict(
+        (
+            path + "/" + k,
+            wandb.Image(((batch_to_canvas(v) + 1) * 127.5).astype(np.uint8)),
```
Comment (jhaux) [inline edflow/hooks/logging_hooks/wandb_handler.py]: This can also be done using 
```
adjust_support(batch_to_canvas, `0->255`, `-1->1`)
```

File: edflow/hooks/logging_hooks/tensorboardX_handler.py [hunk 4]
```diff
@@ -15,7 +15,7 @@ def log_tensorboard_images(writer, results, step, path):
     results = dict((path + "/" + k, v) for k, v in results.items())
     for k, v in results.items():
         v = batch_to_canvas(v)
-        v = ((v+1)*127.5).astype(np.uint8)
+        v = ((v + 1) * 127.5).astype(np.uint8)
```
Comment (jhaux) [inline edflow/hooks/logging_hooks/tensorboardX_handler.py]: This can also be done using
```
adjust_support(batch_to_canvas, `0->255`, `-1->1`)
```

**Review by jhaux — APPROVED** [review]
There are minor changes that would be nice to implement (i.e. `adjust_support` in favor of writing the same functionality explicitly).
In any case this PR is great and will improve working with edflow a lot!
